lib/flex_algo.h

```diff
@@ -0,0 +1,253 @@
+	case FLEX_ALGO_METRIC_TYPE_TE_DEFAULT:
+		return FLEX_ALGO_METRIC_TYPE_TE_DEFAULT_STR;
+	case FLEX_ALGO_METRIC_TYPE_MAX:
+	case FLEX_ALGO_METRIC_TYPE_NONE:
+		break;
+	}
+
+	return "Invalid-Metric-Type";
+}
+
+/*
+ * Defined values for Calculation-Types
+ */
+enum flex_algo_calc_type {
+	FLEX_ALGO_CALC_TYPE_MIN = 0,
+	FLEX_ALGO_CALC_TYPE_SPF = FLEX_ALGO_CALC_TYPE_MIN,	/* 0 */
+	FLEX_ALGO_CALC_TYPE_STRICT_SPF,				/* 1 */
+	FLEX_ALGO_CALC_TYPE_MAX,
+	FLEX_ALGO_CALC_TYPE_NONE
+};
+#define FLEX_ALGO_CALC_TYPE_SPF_STR "spf"
+#define FLEX_ALGO_CALC_TYPE_STRICT_SPF_STR "strict-spf"
+
+static inline enum flex_algo_calc_type flex_algo_str2calc_type(
+	const char *str)
+{
+	int len;
+
+	if (str == NULL)
+		return FLEX_ALGO_CALC_TYPE_NONE;
+
+	len = strlen(str);
+
+	if (strncmp(str, FLEX_ALGO_CALC_TYPE_SPF_STR, len) == 0)
+		return FLEX_ALGO_CALC_TYPE_SPF;
+	else if (strncmp(str, FLEX_ALGO_CALC_TYPE_STRICT_SPF_STR, len) == 0)
+		return FLEX_ALGO_CALC_TYPE_STRICT_SPF;
+
+	return FLEX_ALGO_CALC_TYPE_NONE;
+}
+
+/*
+ * Inline utility to convert metric-types to string.
+ */
+static inline const char *flex_algo_calc_type2str(
+	enum flex_algo_calc_type calc_type)
+{
+	switch (calc_type) {
+	case FLEX_ALGO_CALC_TYPE_SPF:
+		return FLEX_ALGO_CALC_TYPE_SPF_STR;
+	case FLEX_ALGO_CALC_TYPE_STRICT_SPF:
+		return FLEX_ALGO_CALC_TYPE_STRICT_SPF_STR;
+	case FLEX_ALGO_CALC_TYPE_MAX:
+	case FLEX_ALGO_CALC_TYPE_NONE:
+		break;
+	}
+
+	return "Invalid-Calculation-Type";
+}
+
+/*
+ * 4-byte AdminGroup Values
+ */
+PREDECL_LIST(flxalg_admin_groups);
+struct flex_algo_admin_grp {
+	uint32_t admin_grp;
+	struct flxalg_admin_groups_item linkage;
+};
+DECLARE_LIST(flxalg_admin_groups, struct flex_algo_admin_grp, linkage);
+#define FOREACH_FLEX_ALGO_ADMIN_GROUP(list, admingrp)			\
+	frr_each_safe(flxalg_admin_groups, (list), (admingrp))
+
+/*
+ * 4-byte SRLG Values
+ */
+PREDECL_LIST(flxalg_srlgs);
+struct flex_algo_srlg {
+	uint32_t srlg;
+	struct flxalg_srlgs_item linkage;
+};
+DECLARE_LIST(flxalg_srlgs, struct flex_algo_srlg, linkage);
+#define FOREACH_FLEX_ALGO_SRLG(list, srlg)				\
+	frr_each_safe(flxalg_srlgs, (list), (srlg))
+
+/*
+ * 4-byte Flags Values
+ */
+PREDECL_LIST(flxalg_flags);
+struct flex_algo_flags {
+	uint32_t flags;
+	struct flxalg_flags_item linkage;
+};
+DECLARE_LIST(flxalg_flags, struct flex_algo_flags, linkage);
+#define FOREACH_FLEX_ALGO_FLAGS(list, flags)				\
+	frr_each_safe(flxalg_flags, (list), (flags))
+
+/*
+ * The actual Flexible-Algorithm Definition.
+ */
+struct flex_algo_defn {
+	uint8_t id;		/* The Agorithm-Id */
+	uint8_t priority;	/* Priority associated */
+	enum flex_algo_metric_type metric_type; /* Metric-Type associated */
+	enum flex_algo_calc_type calc_type; /* Calculation-Type associated */
+	uint32_t prefix_adv_metric; /* Prefix metric for FAPM sub-TLV */
+	bool prefix_adv_metric_set; /* Prefix metric set or not */
+	/* List of Admin-groups in FA-Exclude-AdminGroup sub-TLV */
+	struct flxalg_admin_groups_head admngrps_exclude;
```

Comment:
Yes you are correct. Will change it to bitfields then..  I have always seen admin groups as 32 bit numbers. :(

Reply:
However I see a issue in packing the admin colours 60, 70, and 4294967295. The bitmap length will be huge. If we would encoded it as 32-bit numbers, it would need only 12 bytes. Anyways that is spec issue and should perhaps discuss in the specs forum.